[Feature:Submission] Add autograder metrics and leaderboard

elihschiff opened the pull request:
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
<!-- List issue if it fixes/closes/implements one using the "Fixes #<number>" or "Closes #<number>" syntax -->
No leaderboard option for gradeables  :sob: 

### What is the new behavior?

This PR adds an optional leaderboard for gradeables. This feature is being added so we can run the HW6 competition in data structures this semester. The leaderboard is set in a gradeable config by adding a leaderboard section, as well as tagging each testcase with the leaderboards it should be a part of. Tagging allows there to be more than one leaderboard, for example a leaderboard for single solution puzzles, a leaderboard for multi solution puzzles, and a leaderboard for both puzzles combined. 

Leaderboards allow for students to de anonymize themselves if they wish to compete. (by default they are anonymous)
Only students in the top X will be visible. The goal is not to make struggling students to feel bad, but instead to allow top students to compete. For students not in the top X, they will see their own score as well as the 3 people (always anonymous) directly ahead of them. They will not see the rank of themselves or the people ahead of them. This method is very common in video games where you want to see the top people as well as your current ranking so you know how much you need to improve to move up.

In order to limit people cheating the ranks by making a fast program that gets partial credit, in order for a testcase to be added to your total points you must have full credit. One can imagine a situation where you might be able to get 80% on a testcase by printing the input back out which would be very fast but would not be fair to other students who did the challenge correctly. The leaderboard is therefore ranked based on the sum of points for testcases where you got full credit, then by the sum of time for those testcases, followed by the sum of memory for those testcases.

To view a sample leaderboard go to http://localhost:1511/courses/f21/sample/gradeable/leaderboard/leaderboard
you will have to add that new gradeable, I am not sure the exact method to do that though.

What students see. In this example the first 2 leaderboards have a limit of 10, the last one is set to 0 meaning it shows all students with some points for that leaderboard.
![student2](https://user-images.githubusercontent.com/18558130/136713087-6c2a6365-ee04-4514-ac05-d2db336741ee.gif)


What instructors see. They see the full leaderboard and the user_id for everyone in the leaderboard. Note that students who have no passing testcases for a leaderboard will not show up.
![instructor](https://user-images.githubusercontent.com/18558130/136712797-fa32d3fd-8ed3-48a3-9ecb-48654d7fa43a.gif)


This PR also adds the ability to track metrics for both time and memory usage for each testcase. That means as an added bonus, we can now display more accurate results next to a testcase. This allows students to really get a sense of how their code is running which can be very helpful when debugging why code is working on your computer but not submitty.  In RPI data structures we often get questions along these lines. In addition this semester in Large Scale with Professor Goldschmidt, we also had many students who would have greatly benefited from this added info. I am sure most classes run into issues like this, but those are 2 examples I know for a fact could benefit. 
FYI for metrics to show up for a student, the submission must be regraded after this PR is applied. The data did not get saved anywhere before this PR.
![image](https://user-images.githubusercontent.com/18558130/136712701-3b03e2c2-d58e-4ed6-83bd-139277dc2f31.png)


### Other information?
<!-- Is this a breaking change? -->
<!-- How did you test -->
This is meant to be a MVP, we are looking for feedback so please feel free to share!

This code was written in partnership with @benjaminrsherman

elihschiff commented:
@williamjallen I have addressed all the comments you have made in the code review. If they are marked as resolved I have fixed them, otherwise I left a response.

In addition for everyone looking at this PR, I have fixed all the comments brought up in the Submitty meeting on Monday. In no specific order we have done:
Added optional description to leaderboards
![image](https://user-images.githubusercontent.com/18558130/137251985-cd796cee-5220-4e75-9850-078c5b913489.png)

Moved the button off the gradeables page to the specific gradeable page

Made the extra info for time/memory on gradeables optional and off by default. Can be turned on by setting `display_testcase_runtime_memory` to `true` in a config.json which one would most likely want to do on a leaderboard gradeable but can be done on any gradeable.

Made the change annon ids now just say "anonomous"
![image](https://user-images.githubusercontent.com/18558130/137252149-a97d5904-69b9-4ab9-b5f7-74f932d65197.png)

remove course staff from leaderboard ranks (note how the number skips over the course staff member)
![image](https://user-images.githubusercontent.com/18558130/137252169-1c0c8da5-457b-46ac-9e08-be7596e08b3b.png)

Removed the extra accurate code in the auto grader going back to the current system that measures in .1s increments

Feel free to let me know anyone has any other comments

bmcutler commented on migration/migrator/migrations/course/20211006221014_leaderboard.py:
```diff
@@ -0,0 +1,55 @@
+    :param course: Code of course being migrated
+    :type course: str
+    """
+    database.execute("DROP TABLE IF EXISTS autograding_metrics;")
+    database.execute("ALTER TABLE electronic_gradeable_version DROP COLUMN IF EXISTS anonymous_leaderboard")
```
If we have run the contest (with all its expensive grading), but I have to rollback to an earlier version temporarily because of an unrelated bug...  I don't want to delete all of the contest data.  It's not simple or quick to "just regrade" the contest. 

Please refactor to not drop the table. 

Thanks.

elihschiff replied:
I have removed the code for the down migration so that it now does nothing. The code is commented out with the idea that if we ever need it, it is there and can be run manually or uncommented out.

https://github.com/Submitty/Submitty/pull/7216/commits/07e2a4b37e14d1ba165563754acc42341f500fb4

williamjallen reviewed (changes requested):
I've been thinking about the UI some for the last few days and I have a few comments.  These aren't strictly necessary changes per se, and they could be added in a future PR, so it would be ok to merge without them but I have listed them below anyway.

1. To me, this kind of UI where different states of control buttons being selected result in different pages of data being displayed screams tabs.  I think that the buttons should be converted to tabs, as used elsewhere on Submitty.  If tabs are not an option due to concerns about the number of them being too large to display horizontally, this could be switched to something where all of the buttons are connected horizontally to give the controls some feeling of connectedness.  See below for an example of what I mean from GitHub.
<img width="490" alt="Untitled" src="https://user-images.githubusercontent.com/16820599/139586593-74f6c73e-5f71-49af-859d-9abd792a1412.png">
<img width="449" alt="Untitled" src="https://user-images.githubusercontent.com/16820599/139586620-95e166a2-0995-4909-85c7-8e5fefe29dcd.png">

2. There's a lot of data here.  That's a good thing but it can be hard to identify where in the list you are.  I suggest adding some sort of identifying background to the row for the current user.  It doesn't have to be a lot, but perhaps something subtle but eye-catching could be added.  If you did this, the "you" text in the row for the current user could be removed
<img width="888" alt="Untitled" src="https://user-images.githubusercontent.com/16820599/139586758-9f6a35aa-b2ee-452b-b0da-a323370373ce.png">

3. In general, user IDs are denoted with angle brackets (e.g. `<bitdiddle>`) when listed after the first+last name elsewhere in Submitty.  It's not a hard rule and doesn't have to be changed but I personally think that it would be better to put the ID in angle brackets in the instructor view.
<img width="186" alt="Untitled" src="https://user-images.githubusercontent.com/16820599/139591319-91b9c348-b920-4e82-86b3-1fa5b6a8c579.png">

4. When trying to visit the leaderboard for a gradeable with no leaderboard, the page redirects to the gradeable list, not the page for the gradeable.  Perhaps it should redirect to the gradeable page instead?

elihschiff commented:
@williamjallen all 4 suggestions should now be fixed here https://github.com/Submitty/Submitty/pull/7216/commits/2677b68ad6c015dd75f961b465e2c0d6ee66f15a
![image](https://user-images.githubusercontent.com/18558130/139594576-66d70c7a-190e-46d4-b222-7c6250b3c8e0.png)

williamjallen reviewed (commented):
Changes look good.  I like the color palette.

My one remaining comment is that the text and button group are not aligned vertically with each other and the button group is rather close to the horizontal line.  Now that the buttons have been grouped, is the "Leaderboards:" text even necessary now?

elihschiff commented:
@williamjallen how is this? I added more space between the buttons and line. I also removed the unneeded "leaderboard:" text https://github.com/Submitty/Submitty/pull/7216/commits/87d51a854f9ee6316e77bb812494430da2525a65
![image](https://user-images.githubusercontent.com/18558130/139596172-89bcc2a6-c6a2-485b-98e1-1f10fc705beb.png)

williamjallen reviewed (approved):
Looks good.  Good work!  I have not reviewed the autograding code beyond a brief glance but  everything else looks good.

bmcutler commented:
@elihschiff @benjaminrsherman 
Some php linting errors remain
(hmm why weren't they failing 2 days ago, and they failed before I merged master)
